### PR TITLE
remove GraphHopper.setEncodingManager 

### PR DESCRIPTION
**Pull request by karussell**
This is required to allow adding encoded values from profiles. 

In the most cases it makes handling of the GraphHopper class easier as only profiles (and no FlagEncoders) needs to be set.

Additionally this helps to reduce usage of the FlagEncoder class as we could add turn cost parser etc from the profile directly.

Required for #2273

### CODE REVIEW

File: core/src/main/java/com/graphhopper/GraphHopper.java [hunk 1]
```diff
@@ -702,16 +713,20 @@ public boolean load(String graphHopperFolder) {
 
         setGraphHopperLocation(graphHopperFolder);
 
-        if (encodingManager == null)
-            setEncodingManager(EncodingManager.create(encodedValueFactory, flagEncoderFactory, ghLocation));
+        if (encodingManager == null) {
+            StorableProperties properties = EncodingManager.loadProperties(ghLocation);
+            encodingManager = properties == null ?
+                    createEncodingManager()
+                    : EncodingManager.create(encodedValueFactory, flagEncoderFactory, properties);
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/GraphHopper.java:720]: Before a missing EncodingManager meant that it should load it from an existing graph folder. 
Now a missing EncodingManager also means that the EncodingManager should be implicitly created from the specified profiles.

Comment (otbutz) [reply core/src/main/java/com/graphhopper/GraphHopper.java:720]: This only creates the EncodingManager from the existing graph if it hasn't been set earlier. This is problematic for things like #2243 where we might have stored additional information in the graph which are not present in the EntityManager being passed in. e.g the encountered String values in this case

I suggest that we always create an EntityManager from the graph if present and throw an exception if their is a mismatch.

File: core/src/main/java/com/graphhopper/routing/util/EncodingManager.java [hunk 2]
```diff
@@ -58,31 +59,25 @@
...
-        return create(new DefaultFlagEncoderFactory(), flagEncodersStr);
+    public static EncodingManager create(String flagEncoders) {
+        return create(new DefaultFlagEncoderFactory(), Arrays.asList(flagEncoders.split(",")));
     }
 
-    public static EncodingManager create(FlagEncoderFactory factory, String flagEncodersStr) {
-        return createBuilder(parseEncoderString(factory, flagEncodersStr)).build();
+    public static EncodingManager create(FlagEncoderFactory factory, List<String> flagEncoderStrings) {
+        return createBuilder(flagEncoderStrings.stream().map(s -> parseEncoderString(factory, s)).collect(Collectors.toList())).build();
     }
```
Comment (easbar) [inline core/src/main/java/com/graphhopper/routing/util/EncodingManager.java:70]: Previously we had a constructor function that took an encoder string and one that took an encoder string and a factory. But now we have one that takes an encoder string and one that takes a list of strings and a factory?

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 1 above)
Comment (easbar) [reply core/src/main/java/com/graphhopper/GraphHopper.java:720]: @karussell  Maybe this code becomes more clear if we move the check for `graph.encoded_values` and `graph.flag_encoders` properties from loadProperties to here?

@otbutz I thought the point is that we cannot set the EncodingManager anymore?

Comment (otbutz) [reply core/src/main/java/com/graphhopper/GraphHopper.java:720]: The EncodingManager can still be populated via `init(GraphHopperConfig)` which is used to create it from the GraphHopperConfig:

https://github.com/graphhopper/graphhopper/blob/785dac682da211411f648849d3d611ac1f3a96f8/core/src/main/java/com/graphhopper/GraphHopper.java#L457-L460

File: core/src/main/java/com/graphhopper/GraphHopper.java [hunk 3]
```diff
@@ -702,16 +713,20 @@ public boolean load(String graphHopperFolder) {
...
+        }
 
         if (!allowWrites && dataAccessType.isMMap())
             dataAccessType = DAType.MMAP_RO;
 
         GHDirectory dir = new GHDirectory(ghLocation, dataAccessType);
         ghStorage = new GraphHopperStorage(dir, encodingManager, hasElevation(), encodingManager.needsTurnCostsSupport(), defaultSegmentSize);
 
-        checkProfilesConsistency();
+        checkProfilesConsistency(encodingManager);
```
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/GraphHopper.java:729]: Not necessary. But it makes it more obvious that the method has always a non-null EncodingManager (?)

File: core/src/main/java/com/graphhopper/storage/GraphBuilder.java [hunk 4]
```diff
@@ -97,11 +97,6 @@ public GraphBuilder setRAM(String location, boolean store) {
         return setDir(new RAMDirectory(location, store));
     }
 
-    public GraphBuilder setBytes(long bytes) {
```
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/storage/GraphBuilder.java:100]: It is only the initial size of the graph and it was not properly used in the cases I saw (like initializing it with the node count or similar). But I can revert this as this is not related.

File: reader-osm/src/test/java/com/graphhopper/GraphHopperTest.java [hunk 5]
```diff
@@ -1742,11 +1781,13 @@ public void testTurnCostsOnOff() {
         final String profile2 = "profile_turn_costs";
         final String vehicle = "car";
         final String weighting = "fastest";
-        GraphHopper hopper = createGraphHopper("car|turn_costs=true").
+        GraphHopper hopper = new GraphHopper().
+                setGraphHopperLocation(GH_LOCATION).
                 setOSMFile(MOSCOW).
+                // add profile with turn costs first when no flag encoder is explicitly added
```
[the remark this answers is not in the document]
Comment (karussell) [reply reader-osm/src/test/java/com/graphhopper/GraphHopperTest.java:1787]: It can be done without this requirement but the check in EncodingManager will be more complicated. E.g. loop through all profiles and add the vehicles for profiles with turn costs first, then without turn costs later.

File: example/src/main/java/com/graphhopper/example/RoutingExampleTC.java [hunk 6]
```diff
@@ -68,10 +68,6 @@ static GraphHopper createGraphHopperInstance(String ghLoc) {
         GraphHopper hopper = new GraphHopper();
         hopper.setOSMFile(ghLoc);
         hopper.setGraphHopperLocation("target/routing-tc-graph-cache");
-        // by enabling turn costs for the FlagEncoder, turn restriction constraints like 'no_left_turn' will be taken
-        // from OSM
```
[the remark this answers is not in the document]
Comment (karussell) [reply example/src/main/java/com/graphhopper/example/RoutingExampleTC.java:72]: Sure

File: core/src/test/java/com/graphhopper/GraphHopperProfileTest.java [hunk 7]
```diff
@@ -66,10 +67,14 @@ public void run() {
...
-    @Test
     public void vehicleDoesNotExist_error() {
-        final GraphHopper hopper = createHopper(EncodingManager.create("car"));
-        hopper.setProfiles(new Profile("profile").setVehicle("your_car"));
+        final GraphHopper hopper = new GraphHopper() {
+            @Override
+            protected void customizeEncodingManager(EncodingManager.Builder emBuilder) {
+                emBuilder.add(new CarFlagEncoder());
+            }
+        }.setGraphHopperLocation(GH_LOCATION).setStoreOnFlush(false).
```
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/test/java/com/graphhopper/GraphHopperProfileTest.java:76]: Not sure as it can still happen with an incorrectly customized EncodingManager.

File: core/src/main/java/com/graphhopper/GraphHopper.java [hunk 8]
```diff
@@ -906,7 +921,10 @@ protected void postProcessing(boolean closeEarly) {
         }
     }
 
-    protected void registerCustomEncodedValues(EncodingManager.Builder emBuilder) {
+    /**
+     * Can be used to newly register encoded values and TagParsers
+     */
+    protected void customizeEncodingManager(EncodingManager.Builder emBuilder) {
```
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/GraphHopper.java:927]: Good questions. This method was there before, I just renamed it and used it more. One different idea would be to give users access to the EncodingManager.Builder and then we are maybe able to remove this method.

File: core/src/main/java/com/graphhopper/routing/util/EncodingManager.java (hunk 2 above)
Comment (karussell) [reply core/src/main/java/com/graphhopper/routing/util/EncodingManager.java:70]: I'm not happy with these methods to be honest. I'll revisit how to make it more consistent.

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 1 above)
Comment (easbar) [reply core/src/main/java/com/graphhopper/GraphHopper.java:720]: Ah ok I see. Yes this way we can set the encoding manager.

>  This is problematic for things like #2243 where we might have stored additional information in the graph which are not present in the EntityManager being passed in. e.g the encountered String values in this case

What exactly is the problem here? Aren't we supposed to use an EncodingManager that is compatible with the stored information?

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (easbar) [reply core/src/main/java/com/graphhopper/GraphHopper.java:729]: Ok I see, it makes it clear that the method depends on encodingManager and this is more critical than e.g. profilesByName, because encodingManager is not final. Hm but it also suggests that the method can/should be used with different encoding managers.. Well, this is probably a matter of taste. The real issue is probably that GraphHopper loads itself and EncodingManager can be null in the first place, but this is out of scope for this PR :)

File: core/src/main/java/com/graphhopper/storage/GraphBuilder.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (easbar) [reply core/src/main/java/com/graphhopper/storage/GraphBuilder.java:100]: Ok either way seems fine. Setting the initial size of the graph can probably be useful, but we use Graphbuilder mostly in tests where the initial size does not matter. But does it ever matter? Even when the storage is resized this probably only takes a fraction of the time of everything else (like reading/importing OSM, running CH etc.)?

File: reader-osm/src/test/java/com/graphhopper/GraphHopperTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (easbar) [reply reader-osm/src/test/java/com/graphhopper/GraphHopperTest.java:1787]: Hm, but this seems dangerous then. It is too easy to add the profiles in the wrong order? I think either we should handle this in GraphHopper, or at least throw an error in case this constraint is violated.

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 1 above)
Comment (karussell) [reply core/src/main/java/com/graphhopper/GraphHopper.java:720]: > This is problematic for things like #2243 where we might have stored additional information in the graph which are not present in the EntityManager being passed in. e.g the encountered String values in this case

I do not see how #2243 would be affected from this PR. Can you explain. To my knowledge all use cases are still possible as before.

File: core/src/test/java/com/graphhopper/GraphHopperProfileTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (easbar) [reply core/src/test/java/com/graphhopper/GraphHopperProfileTest.java:76]: Hm but in then I do not understand the error messages. In both these tests we add a profile for a non-existing vehicle (`your_car`) and this should always give the same error no matter if we added an (unrelated) encoder like CarFlagEncoder?

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 1 above)
Comment (otbutz) [reply core/src/main/java/com/graphhopper/GraphHopper.java:720]: See https://github.com/graphhopper/graphhopper/pull/2243#issuecomment-792769503

tl;dr: the solution proposed in #2243 is currently not working because the EntityManager isn't created from the graph if it was set before by `init()`.

> To my knowledge all use cases are still possible as before.

Your PR doesn't change this behavior but i think it makes sense to change it if we restructure EM handling.

Comment (otbutz) [reply core/src/main/java/com/graphhopper/GraphHopper.java:720]: > Ah ok I see. Yes this way we can set the encoding manager.
> 
> > This is problematic for things like #2243 where we might have stored additional information in the graph which are not present in the EntityManager being passed in. e.g the encountered String values in this case
> 
> What exactly is the problem here? Aren't we supposed to use an EncodingManager that is compatible with the stored information?

The current implementation of the StringEncodedValue only stores **encountered** values in order to safe space. The problem with this approach is that we need to ensure that the EncodingManager is created from the graph in order to retrieve the list of encountered values.

For compatibility checks it's enough to compare the usual EV properties like number of bits etc.

Comment (easbar) [reply core/src/main/java/com/graphhopper/GraphHopper.java:720]: Yes it would be better to load the EncodingManager from the graph. Unfortunately we do not do this yet and this is probably a bigger endeavour. The intention of this PR is that there will be an `is_subnetwork` encoded value that we need per profile (not per encoder). Since there is already `GraphHopper#setProfiles` it would be redundant to also keep `GraphHopper#setEncodingManager` so the idea was to remove `setEncodingManager` and create the encoding manager based on the given profiles. But this certainly won't be the end of the story and we will do more refactoring of the GraphHopper class.

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/GraphHopper.java:729]: Will revert it. We could avoid this problem if we build the EncodingManager private before it is used and do not expose it via a setter. But the getEncodingManager is still used where it is hard to remove it.

[the remark this answers is not in the document]
Comment (easbar) [reply core/src/main/java/com/graphhopper/GraphHopper.java:729]: Yes sure this is out of scope here. I think in the long run there should be the following *functions*

* config.yml -> GraphHopperConfig
* GraphHopperConfig -> New Import / -gh Folder and maybe an in-memory version of this like GraphHopperConfig -> GraphHopper 
* -gh Folder -> GraphHopper
*  GraphHopperBuilder -> GraphHopper for advanced configurations via Java API that cannot be realized by a config object only.

In any case there should be no class that gets somehow configured, leaves some fields null and then 'loads itself'

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 1 above)
Comment (otbutz) [reply core/src/main/java/com/graphhopper/GraphHopper.java:720]: > Unfortunately we do not do this yet and this is probably a bigger endeavour. 

Is it? Wouldn't the following be enough to make it work:

```java
StorableProperties properties = EncodingManager.loadProperties(ghLocation);
if (properties != null) {
     EncodingManager graphEncodingManager = EncodingManager.create(encodedValueFactory, flagEncoderFactory, properties);
     if (encodingManager != null && graphEncodingManager.equals(encodingManager)) {
          throw ...
     }
     encodingManager = graphEncodingManager;
} else if (encodingManager == null) {
    encodingManager = createEncodingManager();
}
```

File: core/src/main/java/com/graphhopper/routing/util/EncodingManager.java [hunk 9]
```diff
@@ -96,29 +91,34 @@ public static EncodingManager create(List<? extends FlagEncoder> flagEncoders) {
         return builder;
     }
 
-    /**
-     * Create the EncodingManager from the provided GraphHopper location. Throws an
-     * IllegalStateException if it fails. Used if no EncodingManager specified on load.
-     */
-    public static EncodingManager create(EncodedValueFactory evFactory, FlagEncoderFactory flagEncoderFactory, String ghLoc) {
+    public static StorableProperties loadProperties(String ghLoc) {
```
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/routing/util/EncodingManager.java]: Was able to replace it with something simpler

File: core/src/main/java/com/graphhopper/storage/GraphBuilder.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/storage/GraphBuilder.java:100]: Yes, very likely it never matters. This is likely also an artifact from the early days where I tried lots of things to improve raw performance.

File: reader-osm/src/test/java/com/graphhopper/GraphHopperTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (karussell) [reply reader-osm/src/test/java/com/graphhopper/GraphHopperTest.java:1783]: Ok, will try to fix this in GraphHopper

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/GraphHopper.java]: Can you have a look into the latest commit where I removed this hook and exposed the EncodingManager.Builder.

File: core/src/test/java/com/graphhopper/GraphHopperProfileTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/test/java/com/graphhopper/GraphHopperProfileTest.java]: The test was ignored (due to missing `@Test`) and indeed the error message is now the same:

> entry in encoder list not supported: your_car

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (easbar) [reply core/src/main/java/com/graphhopper/GraphHopper.java]: Nice, good idea :) This looks much better. GraphHopper in fact *is* partly a builder for itself so it seems to make a lot of sense to be able to use the builder for the encoding manager at least (which builds a *part* of GraphHopper).

[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/GraphHopper.java]: We could also get rid of the XYFactories in most of the use cases as the one who uses GraphHopper now also access to the builder which is used regardless of whether it loads from a `gh` folder or from the config.

File: core/src/main/java/com/graphhopper/routing/util/EncodingManager.java (hunk 2 above)
Comment (easbar) [reply core/src/main/java/com/graphhopper/routing/util/EncodingManager.java:70]: Sure, that sounds useful especially if GraphHopper only wraps the functionality of EncodingManager.Builder anyway

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (easbar) [reply core/src/main/java/com/graphhopper/GraphHopper.java]: Sure, that sounds useful especially if GraphHopper only wraps the functionality of EncodingManager.Builder anyway

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 1 above)
Comment (easbar) [reply core/src/main/java/com/graphhopper/GraphHopper.java]: No, I don't think so. For example we do not store information about profiles.

**Review by easbar — APPROVED** [review]
